src/Tools/babylon.stringDictionary.ts

```diff
@@ -0,0 +1,125 @@
+﻿module BABYLON {
+    /**
+     * This class implement a typical dictionary using a string as key and the generic type T as value.
+     * The underlying implemetation relies on an associative array to ensure the best performances.
+     * The value can be anything including 'null' but except 'undefined'
+     */
+    export class StringDictionary<T> {
+        /**
+         * Get a value based from its key
+         * @param key the given key to get the matching value from
+         * @return the value if found, otherwise undefined is returned
+         */
+        public get(key: string): T {
+            var val = this._data[key];
+            if (val !== undefined) {
+                return val;
+            }
+            return undefined;
+        }
+
+        /**
+         * Get a value from its key or add it if it doesn't exist.
+         * This method will ensure you that a given key/data will be present in the dictionary.
+         * @param key the given key to get the matchin value from
+         * @param factory the factory that will create the value if the key is not present in the dictionary.
+         * The factory will only be invoked if there's no data for the given key.
+         * @return the value corresponding to the key.
+         */
+        public getOrAdd(key: string, factory: (key: string) => T): T {

```

Comment:
Question: should we rename it to getOrAddWithFactory and create a getOrAdd(key, value) which seems more straightforward to me